Application structure -- overview page (#509)
holochain/docs-pages: src/pages/build/application-structure.md

---
title: Application Structure
---

::: topic-list
### In this section {data-no-toc}

> * Application Structure (this page)

Reviewer: ```suggestion
```

Seems redundant

<!--
    * [Zomes] --- integrity vs coordinator, how to structure and compile
    * [DNAs] --- what they're used for, how to specify and bundle
    * [hApps] --- headless vs UI-based, how to bundle and distribute
-->
:::

::: intro
There are a few basic units of composability and packaging you'll need to know about when you're structuring your hApp. Each has different purposes, and the way you break up your code makes a difference to how it works in terms of access, privacy, participant responsibilities, and code reuse.
:::

## Zomes, DNAs, and hApps

### Zomes

> The smallest unit in a hApp is called a chromosome or **zome**. It's the actual binary code that runs in Holochain's [WebAssembly](https://webassembly.org/) VM.

Reviewer: ```suggestion
The smallest unit in a hApp is called a **zome** (a play on chromasome) It's the actual binary code that runs in Holochain's [WebAssembly](https://webassembly.org/) VM.
```

!!! info Why WebAssembly?

We chose WebAssembly because:

* A [number of languages](https://github.com/appcypher/awesome-wasm-langs) can already be compiled to WebAssembly, which we hope will mean Holochain can support many languages on the back end in the future (currently we only supply a back-end SDK for Rust.)
* It's small and fast --- it can get compiled to machine code for near-native speed.

> * Holochain is written in Rust, and Rust has an excellent WebAssembly engine called [Wasmer](https://wasmer.io/) that works on all the major operating systems.

Reviewer: ```suggestion
* It provides a secure sandbox to run untrusted code within.
```

There are two kinds of zome:

* An **integrity zome** defines a set of data types --- your application's schema --- and validation rules for operations that create, update, or delete data of those types. If you're used to thinking of apps in terms of [model-view-controller (MVC)](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller), it's like your application's model.
* A **coordinator zome** defines a set of functions for interacting with data and peers. You can think of it as your application's controller.

Zomes are usually created as pairs --- an integrity zome that defines a portion of your schema and a coordinator zome that defines <abbr title="create, read, update, delete">CRUD</abbr> functions for that schema. You don't have to do it this way though; coordinator zomes don't need an integrity zome if they don't manipulate data, or they can depend on multiple integrity zomes, or multiple coordinators can depend on the same integrity zome.

If you mean for your zomes to be reused by other projects, you can share them via a public Git repository or [crates.io](https://crates.io) (tag your crates with `#holochain` so others can find them).

### DNAs

> Multiple zomes are bundled into a **DNA**. When two or more participants install and run a DNA, an isolated peer-to-peer network is created among them to interact and store shared data.

Reviewer: ```suggestion
Zomes are bundled into a **DNA**. When two or more participants install and run a DNA, an isolated peer-to-peer network is created among them to interact and store shared data.
```

**A DNA, and the network created for it, is uniquely defined by its integrity zomes, plus any modifiers.** The hash of the integrity zomes plus modifiers is called the **DNA hash**, and is the unique identifier for the network.

!!! info Why not coordinator zomes?

Coordinator zomes are bundled with a DNA, but they don't contribute to its uniqueness. This lets you hot-swap coordinators as you fix bugs and add features, without causing the DNA to **fork** a new network. The only things that should cause a fork are changes to integrity code --- the 'rules of the game' for the participants.

!!!

> Because each DNA has its own peer network and data store, you can use the DNA concept to come up with creative approaches to privacy and access, separation of responsibilities, or data retention.

Reviewer: ```suggestion
Because each DNA has its own isolated peer network and data store, you can use the DNA concept to come up with creative approaches to privacy and access, separation of responsibilities, or data retention.
```

### hApp

Multiple DNAs come together in a **hApp** (Holochain app). Each DNA fills a named **role** in the hApp, and can be seen as a [microservice](https://en.wikipedia.org/wiki/Microservices).

> Each peer generates its own public/private key pair when they install a hApp, and this becomes their **agent ID** which identifies and authenticates them as a participant in all the networks created for all the DNAs in the hApp. When a DNA is activated, it's bound to this key pair and becomes a DNA instance or **cell**.

Reviewer: ```suggestion
Each peer generates its own public/private key pair when they install a hApp. Their public key acts as their **agent ID** which identifies them as a participant in all the networks created for all the DNAs in the hApp. When a DNA is activated, it's bound to this key pair and becomes a DNA instance or **cell**.
```

The hApp can specify a few provisioning strategies for its DNAs:

* A cell can be instantiated at app installation time.

> * A cell can be prepared but left uninstantiated until it's needed later.

Reviewer: Not yet implemented. I'd comment it out for now.

Author: Wait, `Create` with `deferred = true` is unimplemented? guh

Author: huh, [well whaddya know](https://github.com/holochain/holochain/blob/e6a7cbce107f7f027cba2b873502f8e9749b904f/crates/holochain_types/src/app/app_bundle.rs#L133). So that's why e2e tests for `deferred: true` are passing even when I try to make a zome call

> * A cell can be prepared but left uninstantiated until it's needed later.

Author: ```suggestion
```

> * If an existing cell that matches a given DNA hash exists, it can be reused.

Author: ```suggestion
```

Guess this is also implemented-but-non-functioning.

* A new cell can be **cloned** from an existing DNA at any time _after the hApp is installed_, with an optional limit on the number of clones.

> * A DNA can be installed but never instantiated into a cell --- instead it serves only as a **stem cell**, a template for cloning.

Reviewer: Also not yet implemented

Author: Implemented, but buggy and non-working. I'll remove it.

> A hApp can optionally include a web-based UI that supporting Holochain runtimes <!-- TODO: link --> can serve to the user.

Reviewer: todo?

Author: after I write #489 there'll be a place to link to.
